### PR TITLE
Adding iterable dataset for training

### PR DESCRIPTION
**Pull request by alecgunny**
Addressing #11

### CODE REVIEW

**Comment by alecgunny** [comment]
- [ ] Build tests for `glitch_frac > 0`
- [ ] Build tests for `waveform_frac > 0`
- [ ] Sample start_indices from first half of arrays to always include trigger
- [ ] Build CI for running tests on PR

File: libs/data/bbhnet/data/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+    def _sample_geocent_time(self, num_samples):
+        raise NotImplementedError
+
+    def inject_waveforms(self, background, waveforms):
+        sample_params = {
+            "ra": self._sample_ra(len(background)),
+            "dec": self._sample_dec(len(background)),
+            "psi": self._sample_psi(len(background)),
+            "geocent_time": self._sample_geocent_time(len(background)),
+        }
```
Comment (alecgunny) [inline libs/data/bbhnet/data/dataloader.py:353]: @wbenoit26 any ideas as to what these sampling functions might be? I assume at least one of them is like a uniform [0, pi] sample or something but not familiar enough with the meaning of all the parameters to be sure

File: libs/data/bbhnet/data/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+        for i, ifo in enumerate(["h1", "l1"]):
+            signal = project_raw_gw(
+                waveforms,
+                sample_params,
+                self.waveform_generator,
+                ifo,
+                get_snr=False,
+            )
+            signals[:, i] = signal
+        return background + torch.Tensor(signals).to(self.device)
```
Comment (alecgunny) [inline libs/data/bbhnet/data/dataloader.py]: @wbenoit26 is something like this sufficient? Or will I need to do the SNR reweighting you do [here](https://github.com/ML4GW/BBHNet/blob/9564f1595a13fa243ebc90a97b7c4025c2f6b3b6/libs/injection/bbhnet/injection/injection.py#L295-L300)? Does the fact that I do whitening by the background PSD help things at all?

Note that at this point the waveforms have been windowed into some `kernel_size` window around the trigger time. Will this affect the ifo response calculation?

**Comment by alecgunny** [comment]
Moving from draft to full PR so that we can start review comments on individual lines. Injection code still needs to be completed and needs tests (though not sure what those will look like), but otherwise should be most of the way there

File: libs/data/bbhnet/data/dataloader.py (hunk 1 above)
Comment (wbenoit26) [reply libs/data/bbhnet/data/dataloader.py:353]: We have some freedom to do what we want here, but I think it makes sense to stick with the priors from some of bilby's example prior dictionaries. They have a Uniform(0, 2*pi) prior on RA, a Cosine prior on Dec, and if the black hole is spinning, a Uniform(0, pi) prior on psi (0 if not spinning). We could have bilby do the sampling here too, but it may not be worth the overhead.

Geocent_time has just been a uniform on the duration of the frame, plus the initial time. For my injection code, there's some logic to prevent signals from overlapping.

File: libs/data/bbhnet/data/dataloader.py (hunk 2 above)
Comment (wbenoit26) [reply libs/data/bbhnet/data/dataloader.py]: The SNR scaling will need to happen here as well if we want to have control over the range we're training on, and it might be a good idea to store the SNR regardless. Whitening will change the SNR, but I don't think that matters if all we're looking for is some consistent way to tune signal strength.

That's a good question. I think the answer is no, but I can do some digging to find out if that's true.

Comment (wbenoit26) [reply libs/data/bbhnet/data/dataloader.py]: Based on bilby's source code, the windowing will not affect the ifo response - it's a purely geometric quantity.

One thing that I do want to flag is the [get_detector_response](https://lscsoft.docs.ligo.org/bilby/_modules/bilby/gw/detector/interferometer.html#Interferometer.get_detector_response) function here, which I suppose is what Tri based his original injection code on. Presumably he had a reason that he needed to customize this code rather than just calling the function, but that might not be the case for us anymore.

We have a working injection script, so I don't want to keep iterating over it, but I did want to note that it exists.

Comment (alecgunny) [reply libs/data/bbhnet/data/dataloader.py]: Sounds good, I'll try to get all these fixes in today so we can merge ASAP.

Good catch on the built-in function. Agreed that since what we have works we may as well keep it, but worth keeping in mind down the line.

File: libs/data/bbhnet/data/dataloader.py (hunk 1 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/dataloader.py]: @wbenoit26 I opted to just use bilby's samplers since they were the most straightforward to implement. I may experiment with using scipy's distribution functions (since numpy doesn't support cosine), but this should work for now.

File: libs/data/bbhnet/data/dataloader.py (hunk 2 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/dataloader.py]: SNR reweighting is implemented, and the windowing ends up being a non-factor because I use the whole signal so that the fft calculation during SNR reweighting is more stable, then window it after it's been projected and reweighted

**Comment by alecgunny** [comment]
@wbenoit26 I think everything should be in here by now, with tests for everything as well. The waveform sampling tests could be a bit more complete, but it's worth having a discussion at some point about what functionality we want to trust that `bbhnet.injection` has tested for us, and which we want to verify in downstream libs. In either case, I've noted areas of improvement with TODO's and can open these up as good-first-issues.

File: libs/data/bbhnet/data/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        assert 0 <= waveform_frac <= 1
+        assert 0 <= glitch_frac <= 1
+
+        self.sample_rate = sample_rate
+        self.kernel_size = int(kernel_length * sample_rate)
+        self.batch_size = batch_size
+        self.batches_per_epoch = batches_per_epoch
+        self.device = device
+
+        # load in the background data and build time-domain
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/dataloader.py:182]: I assume these background files will contain real data. How many seconds of data are we using? Are we using all of this data to calculate the asd? Are we using the same asd to whiten all samples? 

Not super relevant to code, but important questions I've had that I figured should document somewhere.

Comment (EthanMarx) [inline libs/data/bbhnet/data/dataloader.py:182]: I assume these background files will contain real data. How many seconds of data are we using? Are we using all of this data to calculate the asd? Are we using the same asd to whiten all samples? 

Not super relevant to code, but important questions I've had that I figured should document somewhere.

File: libs/data/bbhnet/data/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        # create a target tensor, marking all the glitch data as 0.
+        y = torch.zeros((self.batch_size,))
+
+        # replace some of this data with glitches if
+        # we have glitch data to use
+        if self.glitch_sampler is not None:
+            hanford_glitches, livingston_glitches = self.glitch_sampler.sample(
+                self.num_glitches, self.kernel_size
+            )
+
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/dataloader.py:347]: To clarify, this portion is creating "incoherent" glitches. i.e., no data-sample has a glitch in both Hanford and Livingston. 

In the future, might be worth considering adding some fraction of samples w/ glitches in both Hanford and Livingston.

Comment (EthanMarx) [inline libs/data/bbhnet/data/dataloader.py:347]: To clarify, this portion is creating "incoherent" glitches. i.e., no data-sample has a glitch in both Hanford and Livingston. 

In the future, might be worth considering adding some fraction of samples w/ glitches in both Hanford and Livingston.

File: libs/data/bbhnet/data/utils.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        raise ValueError(
+            "Must specify number of samples N when x is 1 dimensional"
+        )
+    elif len(x.shape) == 1:
+        # this is a 1D array, so we'll just be sampling
+        # the N kernels from the existing timeseries
+        if N > (len(x) - size):
+            raise ValueError(
+                "Array of length {} too short to sample {} "
+                "kernels of length {}".format(len(x), size, N)
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:42]: Should the N and size be switched in ValueError message: `.format(len(x), N, size)`

Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:42]: Should the N and size be switched in ValueError message: `.format(len(x), N, size)`

File: libs/data/bbhnet/data/utils.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    """
+
+    if len(x.shape) == 1 and N is None:
+        raise ValueError(
+            "Must specify number of samples N when x is 1 dimensional"
+        )
+    elif len(x.shape) == 1:
+        # this is a 1D array, so we'll just be sampling
+        # the N kernels from the existing timeseries
+        if N > (len(x) - size):
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:39]: So for example, if `N = (len(x) - size)`, then the kernels will be made by shifting over one sample?

Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:39]: So for example, if `N = (len(x) - size)`, then the kernels will be made by shifting over one sample?

File: libs/data/bbhnet/data/utils.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    for i in idx:
+        # in the 1D case, the idx represent start samples
+        # from the timeseries
+        if len(x.shape) == 1:
+            samples.append(x[i : i + size])
+            continue
+
+        # otherwise, i represents an index along x's 0th dimension,
+        # and we sample from within a kernel's length of
+        # the center of each sampled index, where it is assumed
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:77]: i.e.  For signals, the "peak" or "coalescence" can lie anywhere in the kernel?

Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:77]: i.e.  For signals, the "peak" or "coalescence" can lie anywhere in the kernel?

**Review by EthanMarx — APPROVED** [review]
I made a few comments, but mostly left some clarifying questions. Looks really nice excited to see this in action.

File: libs/data/bbhnet/data/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+        self.device = device
+
+        # load in the background data and build time-domain
+        # filters for whitening using them
+        self.hanford_background, hanford_asd, t0 = _load_background(
+            hanford_background, sample_rate, device
+        )
+        self.livingston_background, livingston_asd, t0 = _load_background(
+            livingston_background, sample_rate, device
+        )
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/dataloader.py:189]: I assume these background files will contain real data. How many seconds of data are we using? Are we using all of this data to calculate the asd? Are we using the same asd to whiten all samples? 

Not super relevant to code, but important questions I've had that I figured should document somewhere.

File: libs/data/bbhnet/data/dataloader.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+            if hanford_glitches is not None:
+                X[: len(hanford_glitches), 0] = hanford_glitches
+                idx = len(hanford_glitches)
+            else:
+                idx = 0
+
+            if livingston_glitches is not None:
+                slc = slice(idx, idx + len(livingston_glitches))
+                X[slc, 1] = livingston_glitches
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/dataloader.py:356]: To clarify, this portion is creating "incoherent" glitches. i.e., no data-sample has a glitch in both Hanford and Livingston.

In the future, might be worth considering adding some fraction of samples w/ glitches in both Hanford and Livingston.

File: libs/data/bbhnet/data/utils.py (hunk 5 above)
Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:42]: So for example, if `N = (len(x) - size)`, then the kernels will be made by shifting over one sample?

File: libs/data/bbhnet/data/utils.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        # that the "trigger" of the relevant event will live
+        if max_sample_start <= 0:
+            start = 0
+        else:
+            start = np.random.randint(min_sample_start, max_sample_start)
```
Comment (EthanMarx) [inline libs/data/bbhnet/data/utils.py:82]: i.e.  For signals, the "peak" or "coalescence" can lie anywhere in the kernel?

**Review by EthanMarx — APPROVED** [review]
I made a few comments, but mostly left some clarifying questions. Looks really nice excited to see this in action.

File: libs/data/bbhnet/data/dataloader.py (hunk 8 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/dataloader.py:189]: So I think ideally this implementation should be agnostic to those details, subject to the constraint that it's sufficiently small data to fit into (GPU) memory at once, so O(1GB). Which data and how much of it will change over time, but starting somewhere on the order of ~16K seconds might be ok to get started? Idk this is probably something good to discuss as part of #27 

But yeah the data will be real and all of it will be used to calculate the ASD, which will be used to whiten all samples both during training _and_ at inference time (the nice thing about structuring the whitening in Torch code is that we can save this step as part of the model so that it gets performed on the server at inference time).

File: libs/data/bbhnet/data/dataloader.py (hunk 9 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/dataloader.py:356]: Ah ok yeah this is a very good point. Would it make sense to merge as-is and make sure that we can get training to converge with this set up, but keep an issue open to fix this once we're confident we have things structured the right way?

File: libs/data/bbhnet/data/utils.py (hunk 5 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/utils.py:42]: Yes on both fronts, great catch. I'll fix that now

File: libs/data/bbhnet/data/utils.py (hunk 10 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/utils.py:82]: Yeah it can lie anywhere in the kernel, but it _has_ to end up in the kernel

File: libs/data/bbhnet/data/dataloader.py (hunk 8 above)
Comment (EthanMarx) [reply libs/data/bbhnet/data/dataloader.py:189]: Got it. Just wanted to bring this up because the ASD is quite variable. Whitening an entire training set w/ the same asd may turn out to be problematic, although as a first pass this should be totally fine.

File: libs/data/bbhnet/data/dataloader.py (hunk 9 above)
Comment (EthanMarx) [reply libs/data/bbhnet/data/dataloader.py:356]: Yes completely agree

**Review by EthanMarx — APPROVED** [review]

File: libs/data/bbhnet/data/dataloader.py (hunk 8 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/dataloader.py:189]: Yeah I think the idea in the context of DL is just to use this as a normalization step, and in that case you just want to make sure you're applying the same normalization to test-time data as you did during training so that the statistics stay the same. If the statistics of your input data have shifted so dramatically that this whitening is no longer helpful, that's a problem but it isn't necessarily specific to whitening: you just have to retrain your network because the distribution you optimized for is no longer the one you're sampling from.

We could do an ifo-wise 0-mean unit-variance standardization instead, but this might be more natural in this context? Idk definitely something to keep an eye on
